src/t_set.c

```diff
@@ -714,82 +709,37 @@ void srandmemberWithCountCommand(redisClient *c) {
         return;
     }
 
-    /* For CASE 3 and CASE 4 we need an auxiliary dictionary. */
-    d = dictCreate(&setDictType,NULL);
-
     /* CASE 3:
-     * The number of elements inside the set is not greater than
-     * SRANDMEMBER_SUB_STRATEGY_MUL times the number of requested elements.
-     * In this case we create a set from scratch with all the elements, and
-     * subtract random elements to reach the requested number of elements.
-     *
-     * This is done because if the number of requsted elements is just
-     * a bit less than the number of elements in the set, the natural approach
-     * used into CASE 3 is highly inefficient. */
-    if (count*SRANDMEMBER_SUB_STRATEGY_MUL > size) {
-        setTypeIterator *si;
+     * The number of requested elements is less than the number of
+     * elements inside the set. */
 
-        /* Add all the elements into the temporary dictionary. */
-        si = setTypeInitIterator(set);
-        while((encoding = setTypeNext(si,&ele,&llele)) != -1) {
-            int retval = DICT_ERR;
-
-            if (encoding == REDIS_ENCODING_INTSET) {
-                retval = dictAdd(d,createStringObjectFromLongLong(llele),NULL);
-            } else {
-                retval = dictAdd(d,dupStringObject(ele),NULL);
-            }
-            redisAssert(retval == DICT_OK);
-        }
-        setTypeReleaseIterator(si);
-        redisAssert(dictSize(d) == size);
+    /* We need an auxiliary set. Optimistically, we create a set using an
+     * Intset internally. */
+    objele = createStringObjectFromLongLong(0);
+    aux_set = setTypeCreate(objele);
+    decrRefCount(objele);
 
-        /* Remove random elements to reach the right count. */
-        while(size > count) {
-            dictEntry *de;
+    /* Get the count requested of random elements from the set into our
+     * auxiliary set. */
+    elements_returned = setTypeRandomElements(set, count, aux_set);
```

Comment:
I think this never was a good idea since `setTypeRandomElements()` specifically notes in its top comment that:
_"Note that this function is not suitable when you need a good distribution of the returned items, but only when you need to "sample" a given number of continuous elements to run some kind of algorithm or to produce statistics."_

But in any case `setTypeRandomElements()` was remove in 9feee428f25a5681a06cd13ed1c4cc644759e719.